libr/arch/arch_session.c

```diff
@@ -0,0 +1,152 @@
+/* radare - LGPL - Copyright 2020 - pancake */
+
+#include <r_arch.h>
+#include <r_asm.h>
+#include "../config.h"
+
+
+static bool r_arch_session_set_cpu(RArchSession *ai, const char *cpu) {
+	r_return_val_if_fail (ai && cpu, false);
+	if (ai && *cpu) {
+		// TODO: improve the check with r_str_split_list()
+		if (R_STR_ISEMPTY (ai->cur->cpus) || strstr (ai->cur->cpus, cpu)) {
+			free (ai->setup.cpu);
+			ai->setup.cpu = cpu? strdup (cpu): NULL;
+		}
+		return true;
+	}
+	return false;
+}
+
+static bool r_arch_session_set_syntax(RArchSession *ai, RArchSyntax syntax) {
+	r_return_val_if_fail (ai, false);
+	switch (syntax) {
+	case R_ASM_SYNTAX_REGNUM:
+	case R_ASM_SYNTAX_INTEL:
+	case R_ASM_SYNTAX_MASM:
+	case R_ASM_SYNTAX_ATT:
+	case R_ASM_SYNTAX_JZ:
+		if (ai->cur->syntax & syntax) {
+			ai->setup.syntax = syntax;
+			return true;
+		}
+		break;
+	default:
+		return false;
+	}
+	return false;
+}
+
+static bool r_arch_session_set_endian(RArchSession *ai, RArchEndian endian) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	switch (endian) {
+	case R_SYS_ENDIAN_LITTLE:
+	case R_SYS_ENDIAN_BIG:
+	case R_SYS_ENDIAN_NONE:
+	case R_SYS_ENDIAN_BI:
+		if (ai->cur->endian & endian) {
+			ai->setup.endian = endian;
+			return true;
+		}
+		break;
+	}
+	return false;
+}
+
+static bool has_bits(RArchPlugin *h, RArchBits bits) {
+        return (h && h->bits && (bits & h->bits));
+}
+
+R_API bool r_arch_session_set_bits(RArchSession *ai, RArchBits bits) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	if (has_bits (ai->cur, bits)) {
+		ai->setup.bits = bits;
+		return true;
+	}
+	return false;
+}
+
+R_API bool r_arch_session_can_decode(RArchSession *ai) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	return ai->cur->decode;
+}
+
+R_API bool r_arch_session_can_encode(RArchSession *ai) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	return ai->cur->encode;
+}
+
+R_API bool r_arch_session_encode(RArchSession *ai, RArchInstruction *ins, RArchOptions opt) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	if (ai->cur->encode) {
+		return ai->cur->encode (ai, ins, opt);
+	}
+	return false;
+}
+
+R_API bool r_arch_session_decode(RArchSession *ai, RArchInstruction *ins, RArchOptions opt) {
+	r_return_val_if_fail (ai && ai->cur, false);
+	if (ai->cur->decode) {
+		return ai->cur->decode (ai, ins, opt);
+	}
+	return false;
+}
+
+R_API void r_arch_session_free(RArchSession *as) {
+	// something more
+	free (as);
+}
+
+R_API RArchSession *r_arch_session_new(RArch *a, RArchPlugin *ap, RArchSetup *setup) {
```

Comment:
shouldn't it call `r_arch_session_ref` before returning? Otherwise you may have return a reference that is already ready to be freed. In general, I think, `_new` should set refcount to 1. See https://github.com/radareorg/radare2/pull/16640/files#diff-6923b12d4ceb725eb8d67413f6807fafR106 for example. That test case, IMHO, should work. And it uses `r_arch_session_unref` at the end.